packages/web-ui/src/ds/molecules/Chat/Message/index.tsx

```diff
@@ -58,30 +76,117 @@ export function Message({
   )
 }
 
+type Segment = string | ReactNode
+
 const ContentValue = ({
   index = 0,
   color,
   value,
   size,
+  parameters = {},
+  collapseParameters = false,
+  sourceMap = [],
 }: {
   index?: number
   color: TextColor
   value: string | Uint8Array | Buffer | ArrayBuffer | URL
   size?: 'default' | 'small'
+  parameters?: Record<string, unknown>
+  collapseParameters?: boolean
+  sourceMap?: PromptlSourceRef[]
 }) => {
-  // TODO: Handle the rest of types
-  if (typeof value !== 'string') return
+  if (typeof value !== 'string') {
+    value = value.toString()
+  }
 
   const TextComponent = size === 'small' ? Text.H6 : Text.H5
 
-  return value?.split('\n')?.map((line, lineIndex) => (
+  // Filter source map references that aren't parameters or don't have a value
+  sourceMap = sourceMap.filter(
+    (ref) =>
+      ref.identifier &&
+      !!parameters[ref.identifier] &&
+      value.slice(ref.start, ref.end).trim().length > 0,
+  )
+
+  // Sort source map to ensure references are ordered
+  sourceMap = sourceMap.sort((a, b) => a.start - b.start)
+
+  // Segment message by source map references
+  let segments: Segment[] = []
+  const firstSegment = value.slice(0, sourceMap[0]?.start ?? value.length)
+  if (firstSegment.length > 0) segments.push(firstSegment)
+  for (let i = 0; i < sourceMap.length; i++) {
+    if (collapseParameters) {
+      segments.push(
+        <Tooltip
+          asChild
+          trigger={
+            <Badge variant='accent'>
+              &#123;&#123;{sourceMap[i]!.identifier}&#125;&#125;
+            </Badge>
+          }
+        >
+          <div className='line-clamp-6'>
+            {parameters[sourceMap[i]!.identifier!] as string}
+          </div>
+        </Tooltip>,
+      )
+    } else {
+      segments.push(
+        <Tooltip
+          asChild
+          trigger={
+            <span className={colors.textColors.accentForeground}>
+              {parameters[sourceMap[i]!.identifier!] as string}
+            </span>
+          }
+        >
+          <div className='line-clamp-6'>{sourceMap[i]!.identifier!}</div>
+        </Tooltip>,
+      )
+    }
+    const nextSegment = value.slice(
+      sourceMap[i]!.end,
+      sourceMap[i + 1]?.start ?? value.length,
+    )
+    if (nextSegment.length > 0) segments.push(nextSegment)
+  }
```

Comment:
tell the ai to refactor for readability there's lots of code that can be put in chunks here